src/message/fetchActions.js

```diff
@@ -170,9 +169,9 @@ export const fetchOlder = (narrow: Narrow): ThunkAction<void> => (dispatch, getS
   const firstMessageId = getFirstMessageId(state, narrow);
   const caughtUp = getCaughtUpForNarrow(state, narrow);
   const fetching = getFetchingForNarrow(state, narrow);
-  const { needsInitialFetch } = getSession(state);
+  const { loading } = getSession(state);
 
-  if (!needsInitialFetch && !fetching.older && !caughtUp.older && firstMessageId !== undefined) {
+  if (!loading && !fetching.older && !caughtUp.older && firstMessageId !== undefined) {
```

Comment:
Hmm. What are the situations where this would produce a different result? I feel like I need to think through some more reasoning to be confident this won't result in some behavior that feels buggy.

I definitely like the idea of reducing entanglement with `needsInitialFetch`. Grepping after this change, it looks like… the only remaining place we consume it is in `AppDataFetcher` itself! The other references are all in comments, tests, or where we update it (but never read it) in `sessionReducer`.

Reply:
It looks like we started using `needsInitialFetch` in 7a7a63fb1, with the commit message

```
Prevent extra fetches before initial fetch is complete
```

Reply:
So if `needsInitialFetch` is `true`, then `loading` should be either
- `true` because the initial fetch is in progress, or
- `false`, but about to follow `needsInitialFetch` in switching to `true`, because an initial fetch is about to start (`AppDataFetcher`'s `useEffect` is about to call `doInitialFetch`).

In the second case, I guess it technically matters whether there's a yield between `needsInitialFetch` becoming `true` and `loading` becoming `true`. I think there isn't such a yield, but I could be wrong.

If there is, then I guess it becomes possible that we do a fetch while `needsInitialFetch` is `true` but `loading` is still false. I think that environment exists in the following ways:
- the app has just started up and rehydrated (and we discover we have an active, logged-in account)
- the server has on the server invalidating an existing event queue after 10 minutes of inactivity

I think it's not really realistic for `fetchOlder` or `fetchNewer` to be called in the first case, at startup. They're called from a UI that has to be reached by navigating.

For the second case...I guess (and again this is only if there's a yield) it would mean we pretty minimally extend the existing window during which it's possible for this code to start a message fetch that ends up still being in progress when an initial fetch starts.

Reply:
If `needsInitialFetch` is `false`, then `loading` should be `false` too:

- All of the cases in `sessionReducer` that set `needsInitialFetch` to `false` also set `loading` to false.
- There's one case where `loading` is set to `true` and `needsInitialFetch` isn't touched. This is on a `INITIAL_FETCH_START`, and at that point `needsInitialFetch` will be `true` too (and will stay that way until the initial fetch completes or aborts).

Reply:
> In the second case, I guess it technically matters whether there's a yield between `needsInitialFetch` becoming `true` and `loading` becoming `true`. I think there isn't such a yield, but I could be wrong.

I'm not sure exactly what happens between the reducer updating `needsInitialFetch` to true, and the effect in `AppDataFetcher` being run. There must be a bunch of React Redux machinery. I agree there probably isn't a yield.

Once that effect starts executing, we go quite promptly back to the reducer to update `loading` to true, with no yield, because `dispatch(initialFetchStart());` is the first line of the `doInitialFetch` body.


> * the server has on the server invalidating an existing event queue after 10 minutes of inactivity

Yeah, I think this is the plausible case where this could make any difference.

Looking at `sessionReducer`, there are a couple of other cases that also give us `needsInitialFetch` true while `loading` still false:
* we've just logged in;
* we've just switched accounts.

But just like just starting up and rehydrating, those lead to a UI where you can't trigger a `fetchOlder` or `fetchNewer` without first navigating somewhere.


> If `needsInitialFetch` is `false`, then `loading` should be `false` too:

I agree with this reasoning.

One correction that doesn't change the conclusion: there's a case that can set `needsInitialFetch` to false that doesn't touch `loading`, which is on rehydrate. But `loading` is always false then, being untouched from its initial value.

Reply:
Thanks for going through that logic! I think this reasoning suffices to conclude that this change is OK.

The key points are:
* `needsInitialFetch` is never false when `loading` is true
* whenever `needsInitialFetch` is true and `loading` isn't, `AppDataFetcher` will immediately cause `loading` to become true
* so at most, this means
  > we pretty minimally extend the existing window during which it's possible for this code to start a message fetch that ends up still being in progress when an initial fetch starts.

It's a bug, and a potential source of wrong data, that that window exists -- or rather, that we don't cancel that earlier message fetch and may use its results, which are potentially missing updates that happened before we created our new event queue. But this change doesn't meaningfully expand that bug, and it helps us toward simplifying the `needsInitialFetch` / `AppDataFetcher` logic.